include/nil/crypto3/multiprecision/cpp_int_modular.hpp

```diff
@@ -0,0 +1,807 @@
+#endif
+                                              || boost::multiprecision::literals::detail::is_value_pack<Arg>::value ||
+                                              (is_trivial_cpp_int_modular<Base>::value &&
+                                               boost::multiprecision::detail::is_arithmetic<Arg>::value),
+                                          std::integral_constant<bool, true>,
+                                          std::integral_constant<bool, false>>::type {
+            };
+
+            template<unsigned Bits>
+            class cpp_int_modular_backend
+                : public cpp_int_modular_base<Bits, is_trivial_cpp_int_modular<cpp_int_modular_backend<Bits>>::value> {
+
+            public:
+                using self_type = cpp_int_modular_backend<Bits>;
+                using base_type = cpp_int_modular_base<Bits, is_trivial_cpp_int_modular<self_type>::value>;
+                using trivial_tag = std::integral_constant<bool, is_trivial_cpp_int_modular<self_type>::value>;
+#ifdef TVM
+                using unsigned_types = std::tuple<unsigned, limb_type, double_limb_type>;
+#else
+
+                using unsigned_types = typename std::conditional<is_trivial_cpp_int_modular<self_type>::value,
+                                                                 std::tuple<unsigned char,
+                                                                            unsigned short,
+                                                                            unsigned,
+                                                                            unsigned long,
+                                                                            boost::ulong_long_type,
+                                                                            double_limb_type>,
+                                                                 std::tuple<limb_type, double_limb_type>>::type;
+                using cpp_int_type = boost::multiprecision::cpp_int_backend<
+                    Bits, Bits,  boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked>; 
+#endif
+                inline constexpr cpp_int_modular_backend() noexcept
+                { }
+                inline constexpr cpp_int_modular_backend(const cpp_int_modular_backend& o) noexcept
+                    : base_type(o)
+                { }
+
+                // rvalue copy:
+                inline constexpr cpp_int_modular_backend(cpp_int_modular_backend&& o) noexcept
+                    : base_type(static_cast<base_type&&>(o)) {
+                }
+
+                // Sometimes we need to convert from one bit length to another. For example from 'Backend_doubled_limbs' to 'Backend'.
+                template<unsigned Bits2>
+                inline constexpr cpp_int_modular_backend(
+                        cpp_int_modular_backend<Bits2>&& o) noexcept {
+                    // Call operator=, which will call do_assign.
```

Comment:
Regarding all the comments about lossy conversion from higher bit value to lower one. In Boost.MP it is handled with `Checked` type template param. I think, this could save us a lot of debugging time. We could make type checked in debug, and unchecked in release to avoid ralated slowdowns

Reply:
https://github.com/NilFoundation/crypto3-multiprecision/issues/72